File: linkedlist/components/linkedlist.rb
class LinkedList
  attr_accessor :head, :tail, :length
  def initialize
    @length = 0
    @head = nil
    @tail = nil
  end

  def add(value=nil, name=nil)
    node = LLNode.new(value, name)
    if !@head
      @head = node
      @tail = node
    else
      cur = @tail
      @tail.next = node
      @tail = node
      @tail.previous = cur
    end
    @length += 1
  end

  def add_head(value=nil, name=nil)
    node = LLNode.new(value, name)
    node.next = @head
    iterate
    @head.previous = node
    @head = node
  end

  def remove_index(index)
    return false if @length < index
    current = @head
    (index - 1).times do
      current = current.next
    end
    current.previous.next = current.next if current.previous
    current.next.previous = current.previous if current.next
    iterate(false)
  end

  def remove
    iterate(false)
    current = @head.next
    current.previous = nil
    @head = current
  end

  def remove_tail
    prev = @tail.previous
    prev.next = nil
    @tail = prev
    iterate(false)
  end

  private
  def iterate(un = true)
    @length += un ? 1 : -1
  end
end

File: linkedlist/components/linkedList.js
export default class LinkedList {
  _head = null;
  _tail = null;
  _length = 0;
  len = () => this._length;

  add(node) {
    /*  Function will add a new node to a linked list. */
    const tail = this._tail, head = this._head;
    if(this._head === null) {
      this._head = node;
    } else {
      node._previous = tail;
    }
    if(tail === null) {
      this._tail = node;
    } else {
      tail._next = node;
      node._previous = tail;
      this._tail = node;
    }
    this._length++;
  }

  pop() {
    /*  Function will return the last node in a linked list. */
    let tails = this._tail;
    if( tails ){
      if( tails._previous ){
        let previous = tails._previous;
        this._tail = previous;
        previous.next = null;
      } else {
        this._tail = null;
        this._head = null;
      }
      this._length--;
    } else {
      return false
    }
  }

  search(value) {
    /*  Function will search through a linked list and return true or
        false for whether the value exists in the linked list or not. */
    let head = this._head;
    while(head) {
      if(head._value === value) break;
      head = head._next;
    }
    return Boolean(head);
  }

  remove(value) {
    /*  Function will remove and return a node from a linked list.
        If the node does not exist null should be returned. */
    let head = this._head;
    while(head) {
      if( head._value === value ) {
        if( head._previous ) head._previous._next = head._next;
        if( head._next ) head._next._previous = head._previous;
        this._length--;
        break;
      }
      head = head._next;
    }
    return head;
  }
}
